### ISSUE TITLE
UserWarning: The AccumulateGrad node's stream does not match the stream of the node that produced the incoming gradient

### ISSUE THREAD

**Issue by mseeger** [awslabs/keys_values#88]
**Describe the bug**  
This happens since we determine sample-based metrics, maybe only with replay buffers.

Getting this warning:
```
Running backward pass over 36 rows of cells, 36 layers, using activation checkpointing
0it [00:00, ?it/s]/home/ubuntu/virtenvs/keysvals_new/lib/python3.12/site-packages/torch/autograd/graph.py:865: UserWarning: The AccumulateGrad node's stream does not match the stream of the node that produced the incoming gradient. This may incur unnecessary synchronization and break CUDA graph capture if the AccumulateGrad node's stream is the default stream. This mismatch is caused by an AccumulateGrad node created prior to the current iteration being kept alive. This can happen if the autograd graph is still being kept alive by tensors such as the loss, or if you are using DDP, which will stash a reference to the node. To resolve the mismatch, delete all references to the autograd graph or ensure that DDP initialization is performed under the same stream as subsequent forwards. If the mismatch is intentional, you can use torch.autograd.graph.set_warn_on_accumulate_grad_stream_mismatch(False) to suppress this warning. (Triggered internally at /pytorch/torch/csrc/autograd/input_buffer.cpp:240.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
```

**To reproduce**  

For example:
```
CUDA_VISIBLE_DEVICES="0,1" PYTORCH_ALLOC_CONF=expandable_segments:True KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/neurips_exp/lora/qwen3_4b/debug6/logs"  python3 keys_values/__main__.py finetune_long_lora Qwen/Qwen3-4B-Instruct-2507 --out_dir /home/ubuntu/out/finetune/neurips_exp/lora/qwen3_4b/debug6  --precision bf16-true --verbose some --devices 2 --data Helmet --data.dataset_key nq --data.max_length 64k --data.metadata_dir /home/ubuntu/out/finetune/data --train.save_interval 10 --train.micro_batch_size 2 --eval.interval 10  --data.trainloader_longest_first True --kv_cache.cache_length 32768 --kv_cache.chunk_size 2048 --kv_cache.name h2o-torch-quantized8 --kv_cache.cpu_offload False --grad.layers_per_cell 1 --eval.initial_validation False --eval.use_sample_metric True --optimizer.name AdamW --optimizer.learning_rate 0.0005
```

Unclear whether this is a bug, or whether there are performance penalties.

May be due to an autograd graph being kept alive. Make sure loss variables are deleted, etc.
Why only now?

**Comment by mseeger** [comment]
May be related to #92. But there, this happens without any replay buffers.
